pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -137,28 +141,72 @@ func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
 	return labelNames, warnings, nil
 }
 
-type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+type stringSliceFunc func(context.Context, storage.Querier) ([]string, storage.Warnings, error)
+
+type stringSliceFuncJob struct {
+	querier  storage.Querier
+	tenantID string
+	result   []string
+	warnings storage.Warnings
+}
 
 // mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
-// on a querier. It removes duplicates and sorts the result. It doesn't require
-// the output of the stringSliceFunc to be sorted, as results of LabelValues
-// are not sorted.
-//
-// TODO: Consider running stringSliceFunc calls concurrently
+// on per querier in parallel. It removes duplicates and sorts the result. It
+// doesn't require the output of the stringSliceFunc to be sorted, as results
+// of LabelValues are not sorted.
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var jobs = make([]interface{}, len(m.tenantIDs))
+
+	for pos := range m.tenantIDs {
+		jobs[pos] = &stringSliceFuncJob{
+			querier:  m.queriers[pos],
+			tenantID: m.tenantIDs[pos],
+		}
+	}
+
+	run := func(ctx context.Context, jobIntf interface{}) error {
+		// end early when context has ended
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		job, ok := jobIntf.(*stringSliceFuncJob)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", jobIntf)
+		}
+
+		result, resultWarnings, err := f(ctx, job.querier)
+		if err != nil {
+			return errors.Wrapf(err, `error querying {%s="%s"}`, defaultTenantLabel, job.tenantID)
+		}
+
+		job.result = result
+		job.warnings = make(storage.Warnings, len(resultWarnings))
```

Comment:
nit: this always allocates, even if `len(resultWarnings)` is 0.

(I would suggest to simply store `job.warnings = resultWarnings` here, and do the wrapping in `aggregate warnings and deduplicate string results` below.)

Reply:
Fixed in 833a7f9